### PR TITLE
Generate publish manifest before building files

### PR DESCRIPTION
**Pull request by live1206**
[AB#230339](https://dev.azure.com/ceapex/Engineering/_workitems/edit/230339)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6037)

### CODE REVIEW

**Comment by yufeih** [comment]
Nice job @github-actions

File: docs/specs/output.yml [hunk 1]
```diff
@@ -705,6 +705,7 @@ outputs:
   .publish.json: |
     {
       "files": [
+        {"url": "/a", "source_path": "a/original.xml", "has_error": true },
```
Comment (live1206) [inline docs/specs/output.yml:708]: We should output files of errors(exception) with `has_error` true

**Comment by live1206** [comment]
> ⚠[DevSandbox](https://github.com/MicrosoftDocs/DevSandbox)(00:00:03.5940295, 0 more diff)

Expected diff, the files throw during build due to schema not found, we should add them to `.publish.json` with `has_error` true

File: src/docfx/publish/PublishUrlMapBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+        public IEnumerable<(PublishUrlMapItem, FilePath)> GetPublishOutput()
+        {
+            return _publishUrlMap.Values.SelectMany(x => x);
+        }
+
+        private void Initialize()
+        {
+            var builder = new ListBuilder<(PublishUrlMapItem item, Document file)>();
+            var files = _redirectionProvider.Files.Where(x => x.Origin != FileOrigin.Fallback)
```
[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/publish/PublishUrlMapBuilder.cs:58]: what is the difference?

File: src/docfx/publish/PublishUrlMapBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                         .Concat(_buildScope.GetFiles(ContentType.Page).Where(x => x.Origin != FileOrigin.Fallback))
+                         .Concat(_tocMap.GetFiles());
+            using (Progress.Start("Building publish map"))
+            {
+                ParallelUtility.ForEach(_errorLog, files, file => AddItem(builder, file));
+            }
+
+            // resolve output path conflicts
+            var publishMap = builder.ToList();
+            var groupByOutputPath = publishMap.Where(x => x.item.OutputPath != null).GroupBy(x => x.item.OutputPath, PathUtility.PathComparer);
```
Comment (yufeih) [inline src/docfx/publish/PublishUrlMapBuilder.cs:69]: `outputPath` is marked non-nullable, so ignore the `OutputPath != null` check?

File: src/docfx/build/context/Context.cs [hunk 4]
```diff
@@ -52,6 +52,8 @@ internal sealed class Context : IDisposable
 
         public ContributionProvider ContributionProvider { get; }
 
+        public PublishUrlMap PublishUrlMapBuilder { get; }
```
Comment (yufeih) [inline src/docfx/build/context/Context.cs:55]:
```suggestion
        public PublishUrlMap PublishUrlMap { get; }
```

File: src/docfx/build/page/BuildPage.cs [hunk 5]
```diff
@@ -25,55 +25,42 @@ public static List<Error> Build(Context context, Document file)
...
-                context.MonikerProvider.GetConfigMonikerRange(file.FilePath),
-                file.ContentType,
-                file.Mime);
-
             if (errors.Any(e => e.Level == ErrorLevel.Error))
             {
-                context.PublishModelBuilder.Add(file.FilePath, publishItem);
-                return errors;
+                context.ErrorLog.Write(errors);
+                context.PublishModelBuilder.Add(file.FilePath, null, null, outputPath);
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs:31]: remove now that `PublishUrlMap` is the source of truth?
```suggestion
                ~~context.PublishModelBuilder.Add(file.FilePath, null, null, outputPath);~~
```

File: src/docfx/publish/PublishUrlMapItemComparer.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            else if (x.Monikers.HasMonikers && !y.Monikers.HasMonikers)
+            {
+                return -1;
+            }
+
+            var result = PathUtility.PathComparer.Compare(x.Monikers.MonikerGroup, y.Monikers.MonikerGroup);
+
+            if (result == 0)
+            {
+                if (x.SourcePath.Origin == FileOrigin.Redirection && y.SourcePath.Origin == FileOrigin.Main)
```
Comment (yufeih) [inline src/docfx/publish/PublishUrlMapItemComparer.cs:68]: Will `x.SourcePath.Compare(y.SourcePath)` cover line 68 to line 81?

**Comment by OsmondJiang** [comment]
🎉🎉🎉🎉✨✨✨✨✨

**Pull request merged**